nano/node/bootstrap/bootstrap.cpp

```diff
@@ -1,1547 +1,277 @@
-	idle.clear ();
 }
 
-nano::bootstrap_initiator::bootstrap_initiator (nano::node & node_a) :
-node (node_a),
-stopped (false),
-thread ([this]() {
-	nano::thread_role::set (nano::thread_role::name::bootstrap_initiator);
-	run_bootstrap ();
-})
+void nano::bootstrap_initiator::lazy_requeue (nano::block_hash const & hash_a, nano::block_hash const & previous_a, bool confirmed_a)
 {
+	auto lazy_attempt (current_lazy_attempt ());
+	if (lazy_attempt != nullptr)
+	{
+		lazy_attempt->lazy_requeue (hash_a, previous_a, confirmed_a);
+	}
 }
 
-nano::bootstrap_initiator::~bootstrap_initiator ()
+void nano::bootstrap_initiator::add_observer (std::function<void(bool)> const & observer_a)
 {
-	stop ();
+	nano::lock_guard<std::mutex> lock (observers_mutex);
+	observers.push_back (observer_a);
 }
 
-void nano::bootstrap_initiator::bootstrap (bool force, std::string id_a)
+bool nano::bootstrap_initiator::in_progress ()
 {
-	nano::unique_lock<std::mutex> lock (mutex);
-	if (force && attempt != nullptr)
-	{
-		attempt->stop ();
-		condition.wait (lock, [& attempt = attempt, &stopped = stopped] { return stopped || attempt == nullptr; });
-	}
-	if (!stopped && attempt == nullptr)
-	{
-		node.stats.inc (nano::stat::type::bootstrap, nano::stat::detail::initiate, nano::stat::dir::out);
-		attempt = std::make_shared<nano::bootstrap_attempt> (node.shared (), nano::bootstrap_mode::legacy, id_a);
-		condition.notify_all ();
-	}
+	nano::lock_guard<std::mutex> lock (mutex);
+	return !attempts_list.empty ();
 }
 
-void nano::bootstrap_initiator::bootstrap (nano::endpoint const & endpoint_a, bool add_to_peers, bool frontiers_confirmed, std::string id_a)
+std::shared_ptr<nano::bootstrap_attempt> nano::bootstrap_initiator::find_attempt (nano::bootstrap_mode mode_a)
 {
-	if (add_to_peers)
-	{
-		if (!node.flags.disable_udp)
-		{
-			node.network.udp_channels.insert (nano::transport::map_endpoint_to_v6 (endpoint_a), node.network_params.protocol.protocol_version);
-		}
-		else if (!node.flags.disable_tcp_realtime)
-		{
-			node.network.merge_peer (nano::transport::map_endpoint_to_v6 (endpoint_a));
-		}
-	}
-	nano::unique_lock<std::mutex> lock (mutex);
-	if (!stopped)
+	for (auto & i : attempts_list)
 	{
-		if (attempt != nullptr)
-		{
-			attempt->stop ();
-			condition.wait (lock, [& attempt = attempt, &stopped = stopped] { return stopped || attempt == nullptr; });
-		}
-		node.stats.inc (nano::stat::type::bootstrap, nano::stat::detail::initiate, nano::stat::dir::out);
-		attempt = std::make_shared<nano::bootstrap_attempt> (node.shared (), nano::bootstrap_mode::legacy, id_a);
-		if (frontiers_confirmed)
+		if (i->mode == mode_a)
 		{
-			excluded_peers.remove (nano::transport::map_endpoint_to_tcp (endpoint_a));
-		}
-		if (!excluded_peers.check (nano::transport::map_endpoint_to_tcp (endpoint_a)))
-		{
-			attempt->add_connection (endpoint_a);
+			return i;
 		}
-		attempt->frontiers_confirmed = frontiers_confirmed;
-		condition.notify_all ();
 	}
+	return nullptr;
```

Comment:
Could use `std::find_if` algorithm

Reply:
Is this more symbols? Like
```
auto existing (std::find_if (attempts_list.begin (), attempts_list.end (), [&mode_a](auto const & item_a) { return item_a->mode == mode_a; }));
if (existing != attempts_list.end ()
{
	return *existing;
}
return nullptr;
```

Reply:
Was thinking this::
```
auto existing (std::find_if (attempts_list.cbegin (), attempts_list.cend (), [&mode_a](auto const & item_a) {
    return item_a->mode == mode_a;
}));
return (existing != attempts_list.cend ()) ? *existing : nullptr;
```

I prefer the named algorithm in this case than a raw for loop, but perhaps this usage of it isn't as great. Should have said it was optional :P